cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,214 @@
+// -A gcp-vips-local -d 10.0.0.2/32 -j REDIRECT
+
+const (
+	// FIXME: support multiple interfaces
+	gcpMetadataForwardedIPs = "instance/network-interfaces/0/forwarded-ips/?recursive=true"
+	// name of nat chain for iptables masquerade rules
+	gcpMasqChainName      = "gcp-vips"
+	gcpLocalMasqChainName = "gcp-vips-local"
+)
+
+type gcpHandler struct {
+	vips      []string
+	ip4tables *iptables.IPTables
+	// GCP doesn't support IPv6
+	// ip6tables *iptables.IPTables
+	stopCh chan struct{}
+}
+
+var _ handler = &gcpHandler{}
+
+func newGCPHandler() (handler, error) {
+	ipt4, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(iptablesTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcpHandler{
+		ip4tables: ipt4,
+		stopCh:    make(chan struct{}),
+	}, nil
+}
+
+// onFailure: stop syncing iptables rules but don't flush all of them
+// keep the ones that are still being forwarded to this host.
+func (gcp *gcpHandler) onFailure() error {
+	select {
+	case gcp.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+
+	glog.Info("Syncing iptables VIP rules")
+	return gcp.syncRulesOnce()
+}
+
+// onSuccess: either start routes service, or remove down file
+func (gcp *gcpHandler) onSuccess() error {
+	if err := gcp.syncRulesOnce(); err != nil {
+		return err
+	}
+
+	// bz1930457 delete stale conntrack entries to the apiserver
+	// FIXME: implement IPv6 once GCP supports it
+	for _, vip := range gcp.vips {
+		// sanity check to not flush unexpected IPs
+		// provided by the metadata service
+		ip := net.ParseIP(vip)
+		if ip == nil || !ip.IsGlobalUnicast() {
+			continue
+		}
+		filter := &netlink.ConntrackFilter{}
+		filter.AddIP(netlink.ConntrackNatSrcIP, ip)
+		if _, err := netlink.ConntrackDeleteFilter(netlink.ConntrackTable, netlink.FAMILY_V4, filter); err != nil {
+			glog.V(4).Infof("Error deleting conntrack entries for %s", ip.String())
+		}
+	}
+
+	go gcp.syncRulesUntil()
+	return nil
+
+}
+
+// https://github.com/openshift/installer/blob/master/data/data/gcp/network/lb-private.tf
+// healthy_threshold   = 3
+// unhealthy_threshold = 3
+// check_interval_sec  = 2
+// timeout_sec         = 2
+
+func (gcp *gcpHandler) successThreshold() int {
+	return 1
+}
+
+func (gcp *gcpHandler) failureThreshold() int {
+	// LB = 6 seconds, plus 10 seconds for propagation
+	return 8
+}
+
+// syncRulesOnce syncs ip masquerade rules
+// TODO ipv6
+func (gcp *gcpHandler) syncRulesOnce() error {
+	// sync VIPs from the metadata
+	oldIPs := sets.NewString(gcp.vips...)
+	resp, err := metadata.Get(gcpMetadataForwardedIPs)
+	if err != nil {
+		return err
+	}
+	var s []string
+	err = json.Unmarshal([]byte(resp), &s)
+	if err != nil {
+		return err
+	}
+	newIPs := sets.NewString(s...)
+	staleVips := oldIPs.Difference(newIPs).List()
+	gcp.vips = newIPs.List()
```

Comment:
This isn't quite right.

We need to do two things:
1. for ALL vips, regardless of up or down, ensure there is a rule in POSTROUTING (that is, for external traffic).
2. for **only** the down apiserver vip, delete the rule in OUTPUT, but just for that vip.

This is so any connections from the LB are handled, but connections from the local kubelet go to out to the LB.

Reply:
how can you know which is the apiserver VIP, the metadata returns a list of IPs that are being forwarded to the instance
```
[]string{`curl`, `-H`, `Metadata-Flavor: Google`, `http://169.254.169.254/computeMetadata/v1/instance/network-interfaces/0/forwarded-ips/?recursive=true`, `[10.0.0.2,34.75.177.157,34.75.192.108]` }
```
does it makes sense keeping redirect rules to IPs that are not there?

Reply:
you know the apiserver internal vip because you know the apiserver hostname; the apiserver-watcher looks it up for you.

All other LB vips need to be left alone.

Reply:
you know the apiserver internal vip because you know the apiserver hostname; the apiserver-watcher looks it up for you.

All other LB vips need to be left alone.

Reply:
talked with Casey offline, he explained me the correct behavior, is the vip we passed as parameter the one that this code has to handle.